swift/Workflow/Sources/SubtreeManager.swift

```diff
@@ -190,6 +210,46 @@ extension WorkflowNode.SubtreeManager {
 
 }
 
+
+extension WorkflowNode.SubtreeManager {
+    fileprivate final class SinkContext {
+
+        var validationState: ValidationState
+        enum ValidationState {
+            case preparing
+            case valid(handler: (Output) -> Void)
+            case invalid
+        }
+
+        init() {
+            self.validationState = .preparing
+        }
+
+        func handle(action: AnyWorkflowAction<WorkflowType>) {
+            switch validationState {
```

Comment:
This covers the *majority* of pre-use/after-use cases. There is still a gap of `sink.send` being call immediately after `render` completes (ie: in `screenDidChange`). It *does* fail, since the `Property` backing the rendering asserts that it's been re-entrantly updated, but it's a less than ideal failure message.

I'm open to thoughts on this. From the best of my exploration, it would have to be in `WorkflowHost` *after* updating the `rendering` property telling the `WorkflowNode` "you're ready", which seems like a clumsy API.

Reply:
I actually don't think that's a bad idea at all.

When a tree update happens, renderings flow (down|up) to the root. So from the root we end up calling `render()`:

https://github.com/square/workflow/blob/02a87fc6c4507f7c007f5de0d95fa125c796e32e/swift/Workflow/Sources/WorkflowHost.swift#L56

We can immediately call `didRender()` after the render pass, which can propagate through the entire tree. Each node can then activate its event sinks as needed.

Reply:
That sounds very similar to the way it works in Kotlin. Each iteration of the main loop consists of three(-ish) phases:
1. **Render pass** (non-blocking) – constructs the context for the current iteration, subscribes channels to every new signal/event sink, tears down obsolete ones.
    1.5. **Snapshot** (non-blocking, separate pass than rendering but happens immediately after) – constructs snapshot of the current state tree.
2. **Publish** (possibly blocking if backpressure) – send rendering, snapshot, and optionally the last output to the consumer of the workflow runtime.
3. **"Tick"** (usually blocking, bad name) – selects on all channels and waits for the next "thing" to happen. Only allows one atomic "thing" to happen at once, runs all `WorkflowAction`s all the way up the tree, and returns the root workflow's optional `OutputT` for that iteration (which is then emitted in the subsequent iteration's "Publish" pass).

Reply:
That's good insight, and gives me confidence this is an appropriate extra step to add.

The behavior will/should be (the differentiation is the most of this will by synchronous):
**Render pass**
1) **Prepare** (what you listed as render pass): Construct the context and call `render` on the workflow.
  - Invalidate old sinks, remove obsolete subscriptions, child workflows, workers (that are no longer declared).
  - Gather all the sinks, subscriptions, etc from render, in a "pending" state.
2) **Publish** (still synchronous) - Send the rendering and optional output.
3) **Enable** (synchronous as well) - Enable all of the sinks, subscriptions, workers, etc. The subscriptions and workers are explicitly subscribed on a queue that guarantees that they will not emit synchronously, so for them having a separate "enable" isn't technically necessary, but fits the model nicely. This will walk the tree of workflows again to enable the root and all the children/grandchildren in a synchronous pass.

Reply:
hyper meta nit: even though these are "internal" concerns, it'd be cool if we used the same naming for these passes in both code bases. That would let us better describe the contract in shared docs.

How does this sound: Dave comes up with good names in this PR, and Kotlin catches up once that mini-bikeshed settles.

Reply:
The names don't exist in code. "render" is the same in both already I think, and that's really the only part they share.

Reply:
I'm debating where the best place to document this is. I feel like at least initially I'm going to write up this behavior as doc comments in the code (so that it's accurate). Overall, I have been kicking around the idea of writing (ie: have started and stopped a few times) a "implementation" or "lifecycle" or "internals" doc that codifies all of these details.

Also debated adding an "RFC" section to the docs, specifically talking about issues like this (so they're more discoverable than just github issues), but don't have a strong preference yet.

Reply:
I actually think this would be good method documentation for render(). render is the most central method in the library, and this describes the precise contract of when the method is called, what it builds, and what is allowed to happen to the thing that it builds.

Reply:
> I have been kicking around the idea of writing (ie: have started and stopped a few times) a "implementation" or "lifecycle" or "internals" doc that codifies all of these details.

I had a section in the docs rewrite outline for this, but Ray suggested using it for something else: https://docs.google.com/document/d/1hlJZh72ZQ-Abd9XZRm4MjsULWQE2pCDTHy-WrFA_XiE/edit#heading=h.4l7zcgaeyx6f

Reply:
Updated, adding an `enableRendering` method that must be called before any actions are valid to be handled.

Reply:
supernit: not a fan of `enableRendering` – it doesn't actually *do* anything to the rendering. We're really enabling any captured event sinks.

You could either name it based on what it does: `enableEvents()`, etc.

or name it based on the phase in the render cycle

`didEmitRendering()`

Reply:
Updated to `enableEvents` - you are right, that's a much better name.